core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -0,0 +1,182 @@
+      as.systemActorOf(KafkaConsumerActor.props(settings), name)
+    }
+
+    subscription match {
+      case TopicSubscription(topics) =>
+        consumer ! KafkaConsumerActor.Subscribe(topics, KafkaConsumerActor.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+      case TopicSubscriptionPattern(topics) =>
+        consumer ! KafkaConsumerActor.SubscribePattern(topics, KafkaConsumerActor.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+    }
+
+    self = getStageActor {
+      case (_, Terminated(ref)) if ref == consumer =>
+        failStage(new Exception("Consumer actor terminated"))
+    }
+    self.watch(consumer)
+  }
+
+  val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { tps =>
+    buffer = buffer.enqueue(tps.toList)
+    pump()
+  }
+  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { r =>
+    r.map(subSources.get).foreach(_.foreach(_.shutdown()))
+    subSources --= r
+  }
+
+  val subsourceCancelledCB = getAsyncCallback[TopicPartition] { tp =>
+    subSources -= tp
+    buffer :+= tp
+    pump()
+  }
+
+  val subsourceStartedCB = getAsyncCallback[(TopicPartition, Control)] {
+    case (tp, control) => subSources += (tp -> control)
+  }
+
+  setHandler(shape.out, new OutHandler {
+    override def onPull(): Unit = {
+      pump()
+    }
+    override def onDownstreamFinish(): Unit = {
+      performShutdown()
+    }
+  })
+
+  def createSource(tp: TopicPartition): Source[Msg, NotUsed] = {
+    Source.fromGraph(new SubSourceStage(tp, consumer))
+  }
+
+  @tailrec
+  private def pump(): Unit = {
+    if (buffer.nonEmpty && isAvailable(shape.out)) {
+      val (tp, remains) = buffer.dequeue
+      buffer = remains
+      push(shape.out, (tp, createSource(tp)))
+      pump()
+    }
+  }
+
+  override def performShutdown() = {
+    //todo we should wait for subsources to be shutdown and next shutdown main stage
+    subSources.foreach {
+      case (_, control) => control.shutdown()
+    }
+
+    if (!isClosed(shape.out)) {
+      complete(shape.out)
+    }
+    self.become {
+      case (_, Terminated(ref)) if ref == consumer =>
+        shutdowned()
+        completeStage()
+    }
+    consumer ! KafkaConsumerActor.Stop
+  }
+
+  class SubSourceStage(tp: TopicPartition, consumer: ActorRef) extends GraphStage[SourceShape[Msg]] { stage =>
+    val out = Outlet[Msg]("out")
+    val shape = new SourceShape(out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+      new GraphStageLogic(shape) with PromiseControl {
+        val shape = stage.shape
+        var requested = false
+        var self: StageActor = _
+        var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+        implicit val ec = ActorMaterializer.downcast(materializer).executionContext
+
+        override def preStart(): Unit = {
+          subsourceStartedCB.invoke((tp, this))
```

Comment:
It is callback from parent stage. I have no idea why this should be thread safe to call parent stage without callback. Could you explain?